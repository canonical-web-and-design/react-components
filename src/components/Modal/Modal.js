import { nanoid } from "nanoid";
import PropTypes from "prop-types";
import React, { useRef } from "react";

export const Modal = ({ children, close, title }) => {
  const descriptionId = useRef(nanoid());
  const titleId = useRef(nanoid());
  return (
    <div className="p-modal" onClick={close}>
      <div
        className="p-modal__dialog"
        role="dialog"
        aria-labelledby={titleId.current}
        aria-describedby={descriptionId.current}
        onClick={(evt) => evt.stopPropagation()}
      >
        <header className="p-modal__header">
          <h2 className="p-modal__title" id={titleId.current}>
            {title}
          </h2>
          <button
            className="p-modal__close"
            aria-label="Close active modal"
            onClick={close}
          >
            Close
          </button>
        </header>
        <div id={descriptionId.current}>{children}</div>
      </div>
    </div>
  );
};

Modal.propTypes = {
  children: PropTypes.node.isRequired,
  close: PropTypes.func.isRequired,
  title: PropTypes.string.isRequired,
};

export default Modal;
