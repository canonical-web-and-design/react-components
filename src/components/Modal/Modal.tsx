import { nanoid } from "nanoid";
import { ReactElement, ReactNode, useRef, useEffect } from "react";

export type Props = {
  /**
   * Buttons to display underneath the main modal content.
   */
  buttonRow?: ReactNode | null;
  /**
   * The content of the modal.
   */
  children: ReactNode;
  /**
   * Function to handle closing the modal.
   */
  close?: () => void | null;
  /**
   * The title of the modal.
   */
  title?: ReactNode | null;
};

export const Modal = ({
  buttonRow,
  children,
  close,
  title,
}: Props): ReactElement => {
  const descriptionId = useRef(nanoid());
  const titleId = useRef(nanoid());

  // This useEffect sets up listeners so the panel will close if user clicks
  // anywhere else on the page or hits the escape key
  useEffect(() => {
    const keyDown = (e) => {
      if (e.code === "Escape") {
        // Close panel if Esc keydown detected
        if (close) {
          close();
        }
      }
    };

    // Add listener on document to capture keydown events
    document.addEventListener("keydown", keyDown);
    // return function to be called when unmounted
    return () => {
      document.removeEventListener("keydown", keyDown);
    };
  }, [close]);

  return (
    <div className="p-modal" onClick={close}>
      <section
        className="p-modal__dialog"
        role="dialog"
        aria-modal="true"
        aria-labelledby={titleId.current}
        aria-describedby={descriptionId.current}
        onClick={(evt) => evt.stopPropagation()}
      >
        {!!title && (
          <header className="p-modal__header">
            <h2 className="p-modal__title" id={titleId.current}>
              {title}
            </h2>
            {!!close && (
              <button
                className="p-modal__close"
                aria-label="Close active modal"
                onClick={close}
              >
                Close
              </button>
            )}
          </header>
        )}
        <div id={descriptionId.current}>{children}</div>
        {!!buttonRow && (
          <footer className="p-modal__footer">{buttonRow}</footer>
        )}
      </section>
    </div>
  );
};

export default Modal;
