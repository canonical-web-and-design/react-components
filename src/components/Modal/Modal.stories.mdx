import { ArgsTable, Canvas, Meta, Story } from "@storybook/addon-docs/blocks";

import Modal from "./Modal";

<Meta
  title="Modal"
  component={Modal}
  argTypes={{
    children: {
      control: {
        disable: true,
      },
    },
  }}
/>

export const Template = (args) => {
  return (
    <div style={{ height: "100vh", position: "relative" }}>
      <Modal {...args} />
    </div>
  );
};

### Modal

This is a [React](https://reactjs.org/) component for the Vanilla [Modal](https://docs.vanillaframework.io/patterns/modal/).

The modal component can be used to overlay an area of the screen which can contain a prompt, dialog or interaction.

### Props

<ArgsTable of={Modal} />

<Canvas>
  <Story
    name="Default"
    args={{
      close: () => {},
      title: "Confirm delete",
      children: (
        <>
          <p>
            Are you sure you want to delete user "Simon"? This action is
            permanent and can not be undone.
          </p>
          <hr />
          <div className="u-align--right">
            <button onClick={() => {}} className="u-no-margin--bottom">
              Cancel
            </button>
            <button className="p-button--negative u-no-margin--bottom">
              Delete
            </button>
          </div>
        </>
      ),
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>
