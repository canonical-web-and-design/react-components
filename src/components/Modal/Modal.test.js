import { shallow } from "enzyme";
import React from "react";

import Modal from "./Modal";

describe("Modal ", () => {
  it("renders", () => {
    const wrapper = shallow(
      <Modal close={jest.fn()} title="Confirm delete">
        Are you sure?
      </Modal>
    );
    expect(wrapper).toMatchSnapshot();
  });
});
