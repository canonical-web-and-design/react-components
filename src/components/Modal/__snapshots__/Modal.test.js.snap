// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal  displays any content in the button row 1`] = `
<div
  className="p-modal"
>
  <section
    aria-describedby="Uakgb_J5m9g-0JDMbcJqLJ"
    aria-labelledby="Uakgb_J5m9g-0JDMbcJqLJ"
    className="p-modal__dialog"
    onClick={[Function]}
    role="dialog"
  >
    <div
      id="Uakgb_J5m9g-0JDMbcJqLJ"
    >
      Button tester
    </div>
    <footer
      className="p-modal__footer"
    >
      <div>
        I am not a button
      </div>
    </footer>
  </section>
</div>
`;

exports[`Modal  displays buttons that have been supplied 1`] = `
<div
  className="p-modal"
>
  <section
    aria-describedby="Uakgb_J5m9g-0JDMbcJqLJ"
    aria-labelledby="Uakgb_J5m9g-0JDMbcJqLJ"
    className="p-modal__dialog"
    onClick={[Function]}
    role="dialog"
  >
    <div
      id="Uakgb_J5m9g-0JDMbcJqLJ"
    >
      Button tester
    </div>
    <footer
      className="p-modal__footer"
    >
      <button
        key="1"
      >
        Button 1
      </button>
      <button
        key="2"
      >
        Button 2
      </button>
    </footer>
  </section>
</div>
`;

exports[`Modal  renders 1`] = `
<div
  className="p-modal"
  onClick={[MockFunction]}
>
  <section
    aria-describedby="Uakgb_J5m9g-0JDMbcJqLJ"
    aria-labelledby="Uakgb_J5m9g-0JDMbcJqLJ"
    className="p-modal__dialog"
    onClick={[Function]}
    role="dialog"
  >
    <header
      className="p-modal__header"
    >
      <h2
        className="p-modal__title"
        id="Uakgb_J5m9g-0JDMbcJqLJ"
      >
        Confirm delete
      </h2>
      <button
        aria-label="Close active modal"
        className="p-modal__close"
        onClick={[MockFunction]}
      >
        Close
      </button>
    </header>
    <div
      id="Uakgb_J5m9g-0JDMbcJqLJ"
    >
      Are you sure?
    </div>
  </section>
</div>
`;
